Same error is reported multiple times when using multiprocessing (including performance decrease)

<!--
  Hi there! Thank you for discovering and submitting an issue.

  Before you submit this, make sure that the issue doesn't already exist
  or if it is not closed.

  Is your issue fixed on the preview release?: pip install pylint astroid --pre -U

-->

### Steps to reproduce
1. create a new virtualenv with latest pylint development state (3fffdf03d5a32e4a0d1b5216ea9d16dacacdaa6f)
2. create a folder "test" somewhere
3. add an (empty) "__init__.py" file there
4. add an (empty "spam1.py" file there, write "asdf" and nothing else in that file
5. copy the file 20 times in this same folder, calling the files "spam2.py" etc.
6. go to the parent directory of "test" and call "pylint -j 1 -E test":
```
(pylint) C:\work\pylint_test>pylint -j 1 -E test
************* Module test.spam1
test\spam1.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam10
test\spam10.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam11
test\spam11.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam12
test\spam12.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam13
test\spam13.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam14
test\spam14.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam15
test\spam15.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam16
test\spam16.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam17
test\spam17.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam18
test\spam18.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam19
test\spam19.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam2
test\spam2.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam20
test\spam20.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam3
test\spam3.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam4
test\spam4.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam5
test\spam5.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam6
test\spam6.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam7
test\spam7.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam8
test\spam8.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam9
test\spam9.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
```
7. go to the parent directory of "test" and call "pylint -j 0 -E test":
```
(pylint) C:\work\pylint_test>pylint -j 0 -E test
************* Module test.spam11
test\spam11.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam14
test\spam14.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam12
test\spam12.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
test\spam11.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam17
test\spam17.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam10
test\spam10.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam1
test\spam1.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
test\spam14.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam18
test\spam18.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam13
test\spam13.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam16
test\spam16.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam15
test\spam15.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
test\spam11.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
test\spam17.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam2
test\spam2.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
test\spam12.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam19
test\spam19.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
test\spam10.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam20
test\spam20.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
test\spam14.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
test\spam18.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam4
test\spam4.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
test\spam1.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam3
test\spam3.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
test\spam13.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam5
test\spam5.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
test\spam15.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam7
test\spam7.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
test\spam11.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
test\spam17.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
test\spam2.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam8
test\spam8.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
test\spam12.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
test\spam19.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam9
test\spam9.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
test\spam16.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
************* Module test.spam6
test\spam6.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)
```
For example `test\spam11.py:1:0: E0602: Undefined variable 'fdd' (undefined-variable)` is reported four times.

### Current behavior
The same error is reported multiple times.

### Expected behavior
Each error is reported only once (and in the report at the correct place).

### pylint --version output
```
pylint 2.6.0-dev1
astroid 2.4.0
Python 3.6.4 (v3.6.4:d48eceb, Dec 19 2017, 06:54:40) [MSC v.1900 64 bit (AMD64)]
```

--- a/doc/whatsnew/2.7.rst
+++ b/doc/whatsnew/2.7.rst
@@ -16,3 +16,7 @@ New checkers
 
 Other Changes
 =============
+
+* Fix bug that lead to duplicate messages when using ``--jobs 2`` or more.
+
+  Close #3584

--- a/pylint/lint/check_parallel.py
+++ b/pylint/lint/check_parallel.py
@@ -69,6 +69,7 @@ def _worker_check_single_file(file_item):
     _worker_linter.check_single_file(name, filepath, modname)
 
     msgs = [_get_new_args(m) for m in _worker_linter.reporter.messages]
+    _worker_linter.reporter.reset()
     return (
         _worker_linter.current_name,
         filepath,

--- a/pylint/reporters/collecting_reporter.py
+++ b/pylint/reporters/collecting_reporter.py
@@ -16,4 +16,7 @@ class CollectingReporter(BaseReporter):
     def handle_message(self, msg):
         self.messages.append(msg)
 
+    def reset(self):
+        self.messages = []
+
     _display = None

--- a/tests/lint/unittest_lint.py
+++ b/tests/lint/unittest_lint.py
@@ -791,7 +791,6 @@ def test_custom_should_analyze_file():
 def test_multiprocessing(jobs):
     """ Check that multiprocessing does not create duplicates.
     """
-    source_dir = os.path.join(REGRTEST_DATA_DIR)
     # For the bug (#3584) to show up we need more than one file with issues
     # per process
     filenames = [
@@ -810,8 +809,8 @@ def test_multiprocessing(jobs):
     linter.set_reporter(reporter)
 
     try:
-        sys.path.append(os.path.dirname(source_dir))
-        linter.check([os.path.join(source_dir, fname) for fname in filenames])
+        sys.path.append(os.path.dirname(REGRTEST_DATA_DIR))
+        linter.check([os.path.join(REGRTEST_DATA_DIR, fname) for fname in filenames])
     finally:
         sys.path.pop()
 

--- a/tests/lint/unittest_lint.py
+++ b/tests/lint/unittest_lint.py
@@ -785,6 +785,40 @@ def test_custom_should_analyze_file():
         assert "invalid syntax" in messages[0]
 
 
+# we do the check with jobs=1 as well, so that we are sure that the duplicates
+# are created by the multiprocessing problem.
+@pytest.mark.parametrize("jobs", [1, 2])
+def test_multiprocessing(jobs):
+    """ Check that multiprocessing does not create duplicates.
+    """
+    source_dir = os.path.join(REGRTEST_DATA_DIR)
+    # For the bug (#3584) to show up we need more than one file with issues
+    # per process
+    filenames = [
+        "special_attr_scope_lookup_crash.py",
+        "syntax_error.py",
+        "unused_variable.py",
+        "wildcard.py",
+        "wrong_import_position.py",
+    ]
+
+    reporter = testutils.TestReporter()
+    linter = PyLinter()
+    linter.config.jobs = jobs
+    linter.config.persistent = 0
+    linter.open()
+    linter.set_reporter(reporter)
+
+    try:
+        sys.path.append(os.path.dirname(source_dir))
+        linter.check([os.path.join(source_dir, fname) for fname in filenames])
+    finally:
+        sys.path.pop()
+
+    messages = reporter.messages
+    assert len(messages) == len(set(messages))
+
+
 def test_filename_with__init__(init_linter):
     # This tracks a regression where a file whose name ends in __init__.py,
     # such as flycheck__init__.py, would accidentally lead to linting the

